src/posthog-surveys.ts

```diff
@@ -261,38 +291,50 @@ export class PostHogSurveys {
         const existingSurveys = this.instance.get_property(SURVEYS)
 
         if (!existingSurveys || forceReload) {
-            this.instance._send_request({
-                url: this.instance.requestRouter.endpointFor(
-                    'api',
-                    `/api/surveys/?token=${this.instance.config.token}`
-                ),
-                method: 'GET',
-                callback: (response) => {
-                    const statusCode = response.statusCode
-                    if (statusCode !== 200 || !response.json) {
-                        logger.error(`Surveys API could not be loaded, status: ${statusCode}`)
-                        return callback([])
-                    }
-                    const surveys = response.json.surveys || []
-
-                    const eventOrActionBasedSurveys = surveys.filter(
-                        (survey: Survey) =>
-                            (survey.conditions?.events &&
-                                survey.conditions?.events?.values &&
-                                survey.conditions?.events?.values?.length > 0) ||
-                            (survey.conditions?.actions &&
-                                survey.conditions?.actions?.values &&
-                                survey.conditions?.actions?.values?.length > 0)
-                    )
-
-                    if (eventOrActionBasedSurveys.length > 0) {
-                        this._surveyEventReceiver?.register(eventOrActionBasedSurveys)
-                    }
+            // Prevent concurrent API calls
+            if (this._isFetchingSurveys) {
+                return
+            }
 
-                    this.instance.persistence?.register({ [SURVEYS]: surveys })
-                    return callback(surveys)
-                },
-            })
+            try {
+                this._isFetchingSurveys = true
+                this.instance._send_request({
+                    url: this.instance.requestRouter.endpointFor(
+                        'api',
+                        `/api/surveys/?token=${this.instance.config.token}`
+                    ),
+                    method: 'GET',
+                    callback: (response) => {
+                        this._isFetchingSurveys = false
```

Comment:
i think this should be done after the callback is executed so L312 and 330, otherwise you still might have race conditions in the `_surveyEventReceiver` and `persistence`

Reply:
Javascript is monothread, I don't see a way for this code to yield control to a separate function in between here and line 330, so this **should** be fine, and actually looks slightly cleaner to me

Reply:
added a 10s timeout to it, and [created tests](https://github.com/PostHog/posthog-js/pull/1747/commits/b50c1724d4e59040e2f1e2b51350390f6d0135c3#diff-04488c5fc8f15661dc5ea1cd4d4be9e1b0cf35cefaa2c3b84c8f5a29fd6e487fR204-R215) to make sure its happening as expected